_posts/2016-10-02-distributed-array-of-commodity-hosting-services.md

```diff
@@ -0,0 +1,184 @@
+on. We need to run 100 instances of our application to handle the expected
+load. We'll call this zone A.
+
+The zone A PaaS offers a Service Level Agreement (SLA) of 95% availability[^3].
+
+<table>
+<tr>
+<th>PaaS cost (%)</th>
+<th>Availability (%)</th>
+</tr>
+<tr>
+<td>100</td>
+<td>95%</td>
+</tr>
+</table>
+
+How can we improve on that?
+
+## Add a thing in front of it
+
+Remember the [fundamental theorem of software engineering][ftse]:
+
+> We can solve any problem by adding a layer of indirection
+
+First put in another layer in front of our PaaS. This layer offers an SLA of
+99.99% availability. This layer routes traffic to our stateless application
+running in the PaaS.
+
+<table>
+<tr>
+<th>PaaS cost (%)</th>
+<th>Availability (%)</th>
+</tr>
+<tr>
+<td>100</td>
+<td>94.99</td>
+</tr>
+</table>
+
+This has actually made things worse. The serial availability equation tells us
+that the availability can be calculated from the product of all of the
+component's availability in the network path. So that would be
+`0.9999 * 0.95 = 94.9905%` availability.
+
+We're also now paying for another thing -- the routing layer.
+
+## Cope with single zone failure
+
+We get a similar PaaS, with similar SLAs. We deploy our app into that, and
+update our routing layer to send traffic to this new zone B as well. Now we
+see some benefits.
+
+We can cope with an outage happening in zone A; the other zone will just soak up
+the traffic. And vice versa.
+
+We can now use the parallel availability equation. It becomes
+
+`1 - [(1 - 0.95) * (1 - 0.95)] = 0.9975`
+
+But we still have the routing layer at 99.99%, so the serial availability
+equation again gives us 99.740025%. By running 2 zones in parallel, we get a big
+boost above the basic 95% SLA that both PaaS providers offer.
+
+It might be that both zones have an outage at the same time, as part of their
+95% SLA. We can make choices around which provider to use to try to avoid that.
+
+For example, Azure, AWS and Google are unlikely to all have problems at the same
+time. If they do, something really bad has happened.
+
+Unfortunately, we've doubled our costs in a naive implementation. We're now
+paying for 100 instances in both zones.
+
+<table>
+<tr>
+<th>PaaS cost (%)</th>
+<th>Availability (%)</th>
+</tr>
+<tr>
+<td>200</td>
+<td>99.74%</td>
+</tr>
+</table>
+
+Elastic scaling could help with that cost aspect. There are risks that
+if other people have architected things in this way, you'll all be trying to
+autoscale in the same zone. You should probably talk to a relevant stakeholder
+about how much they care whether you solve that problem.
+
+## Cope with multiple zone failures
+
+We decide to add some more PaaS providers. We now have zones C, D, and E, and we
+get to make other decisions.
+
+How many zone failures do we need to be resilient against? If the answer is just
+1, then we only need to put 25 instances of our application in each zone,
```

Comment:
So we'd have 125 instances total across 5 zones? Why do we need this number? *

* Wine again